### ISSUE TITLE
Unclear error when incompatible JavaScript included on page (but works in SSR)

### ISSUE THREAD

**Issue by jonhilt** [dotnet/aspnetcore#51010]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

It seems it's now possible to include `<script>` tags in razor components, and indeed this is used to good effect in App.razor.

This appears to open the door an interesting possibility when using SSR, to interpolate the values of C# variables with JavaScript code.

But when you attempt that with Blazor Server as the render mode it errors and results in a fairly vague error.

### Expected Behavior

It seems this approach works in SSR but not in other render modes.

If this is by design then a clearer error message would be useful.

Failing that, guidance in the docs? (To either advise against it, or outline the limitations).

### Steps To Reproduce

Here's an example:

``` html
@page "/BookingComplete"
@attribute [RenderModeServer]

<h3>BookingComplete</h3>

<p>You're in: @OrderEmail</p>

<script>
    let Bob = '@OrderEmail';
    console.log(Bob);
</script>

@code {
    
    [SupplyParameterFromQuery(Name="orderEmail")]
    public string OrderEmail { get; set; }
    
}
```

### Exceptions (if any)

![image](https://github.com/dotnet/aspnetcore/assets/102787/101562b0-eac1-4ade-bfd9-0effdfde106d)

![image](https://github.com/dotnet/aspnetcore/assets/102787/2a436770-d3db-46b4-ab32-d9ee2a23584d)

### .NET Version

8.0.100-rc.1.23455.8

**Comment by SteveSandersonMS** [comment]
Thanks for reporting it. I think this was already fixed in RC2: https://github.com/dotnet/razor/issues/9204

Would you mind trying a daily build of RC2 to see if it resolves your issue?

**Comment by SteveSandersonMS** [comment]
Of course I should also add that generating script by interpolating user-supplied values is something you need to be careful about. Be sure to check the resulting HTML has the user-supplied values correctly JavaScript-string-encoded, otherwise you're simply adding a cross-site scripting (XSS) vulnerability to your site.

As far as I know the framework will JavaScript-string-encode such values by default, but please double-check.

**Comment by jonhilt** [comment]
Thanks Steve, happy to try this with RC2. 

Where can I grab an RC2 build at this point? I only see RTM here https://github.com/dotnet/installer/blob/main/README.md#installers-and-binaries

**Comment by javiercn** [comment]
@jonhilt if you grab one of the RTM ones that will work too

**Comment by jonhilt** [comment]
OK just tested in a new Blazor project using 8.0.100-rtm.23479.1

Looks like I'm seeing the same error.

Here's a repo: https://github.com/jonhilt/BlazorNet8Testing/tree/master/RTMTest

https://github.com/jonhilt/BlazorNet8Testing/blob/master/RTMTest/Components/Pages/BookingComplete.razor

**Comment by SteveSandersonMS** [comment]
Oh I see, thanks for clarifying. We did a compiler change to make this work for the static SSR case, but we still don't support adding `<script>` dynamically during interactive rendering (and that's what you're doing by using `@attribute [RenderModeInteractiveServer]`).

There are several reasons for not supporting this:

1. When the component interactively re-renders, in many common cases, the behavior would be impossible for developers to make sense of. Depending on which parts of the page are re-rendering, the script may or may not re-execute. There's no way for Blazor to know what is the developer's intent, and no syntax for indicating it.
2. If we did want this to work, we'd have to change the client-side rendering logic to buffer all the nodes it's trying to add to a `<script>`, concatenate them together, and then assign them all at once. It wouldn't work to add each text node as we see it, because the `<script>` has to be completely valid atomically.
3. Some kinds of changes would make no sense ever, such as trying to update/re-render just part of the contents of a `<script>` element.

These are all the reasons why we used to have a compiler error if you added a `<script>` to a `.razor` file. We removed that error because it does make sense to do it in pure SSR pages, and the compiler doesn't know up front whether any given component will later be used in an SSR or interactive context.

Ideally we'd add a runtime check to provide a clearer error at runtime if you add `<script>` during interactive rendering, but I suspect it may be too late and too low-priority to do that for .NET 8. However, I'll let the rest of the team see this in case people feel we can and should add such an error for .NET 8. cc @dotnet/aspnet-blazor-eng
